### ISSUE TITLE
[CodeHealth] Streamline delayed execution and historical data management

### ISSUE THREAD

**Issue by red-0ne** [pokt-network/poktroll#931]
## Objective

Reduce code redundancy and complexity by streamlining delayed flow execution and historical data storage and retrieval.

## Origin Document

[Delayed-execution-On-Chain-Indexing](https://www.notion.so/buildwithgrove/Delayed-execution-On-Chain-Indexing-134a36edfff680ce8484ca4672b643fc#143a36edfff68074b881f8c7e8b0f3b9
)
## Goals

- Harmonize delayed flow execution such as `Supplier` unstaking.
- Harmonize historical data indexing such as block hashes used in C&P lifecycle.

## Deliverables

- [ ] A PR implementing Historical Data Indexing and removing custom implementations.
- [ ] A PR implementing Delayed function execution and removing custom implementations.

## Non-goals / Non-deliverables

- Alter business logic implementation.

## General deliverables

- [ ] **Comments**: Add/update TODOs and comments alongside the source code so it is easier to follow.
- [ ] **Testing**: Add new tests (unit and/or E2E) to the test suite.
---

**Creator**: @red-0ne
**Co-Owners**: @bryanchriswhite

**Comment by Olshansk** [comment]
@red-0ne Lmk once the design doc for this is ready. I'll review it & run it by BinaryBuilders + InformalSystems to ensure we're doing work that compliments the cosmos SDK.

**Comment by Olshansk** [comment]
Added the `nice-to-have` label so we can always find it later but closing it out to keep backlog only with important & high priority items.
